crates/holochain/src/core/sys_validate.rs

```diff
@@ -32,23 +39,251 @@ pub const MAX_ENTRY_SIZE: usize = 16_000_000;
 /// fast lookup so they should be small.
 pub const MAX_TAG_SIZE: usize = 1000;
 
+/// Need at least two to countersign.
+pub const MIN_COUNTERSIGNING_AGENTS: usize = 2;
+/// 8 seems like enough agents to countersign.
+pub const MAX_COUNTERSIGNING_AGENTS: usize = 8;
+
 /// Verify the signature for this header
-pub async fn verify_header_signature(
-    sig: &Signature,
-    header: &Header,
-) -> SysValidationResult<bool> {
+pub async fn verify_header_signature(sig: &Signature, header: &Header) -> SysValidationResult<()> {
     if header.author().verify_signature(sig, header).await? {
-        Ok(true)
+        Ok(())
     } else {
-        Ok(false)
+        Err(SysValidationError::ValidationOutcome(
+            ValidationOutcome::Counterfeit((*sig).clone(), (*header).clone()),
+        ))
     }
 }
 
 /// Verify the author key was valid at the time
 /// of signing with dpki
 /// TODO: This is just a stub until we have dpki.
-pub async fn author_key_is_valid(_author: &AgentPubKey) -> SysValidationResult<bool> {
-    Ok(true)
+pub async fn author_key_is_valid(_author: &AgentPubKey) -> SysValidationResult<()> {
+    Ok(())
+}
+
+/// Verify that the signature on a preflight request is valid.
+pub async fn check_countersigning_preflight_response_signature(
+    preflight_response: &PreflightResponse,
+) -> SysValidationResult<()> {
+    let signature_is_valid = preflight_response
+        .request()
+        .signing_agents()
+        .get(*preflight_response.agent_state().agent_index() as usize)
+        .ok_or_else(|| {
+            SysValidationError::ValidationOutcome(ValidationOutcome::PreflightResponseSignature(
+                (*preflight_response).clone(),
+            ))
+        })?
+        .0
+        .verify_signature_raw(
+            preflight_response.signature(),
+            &preflight_response.encode_for_signature().map_err(|_| {
+                SysValidationError::ValidationOutcome(
+                    ValidationOutcome::PreflightResponseSignature((*preflight_response).clone()),
+                )
+            })?,
+        )
+        .await?;
+    if signature_is_valid {
+        Ok(())
+    } else {
+        Err(SysValidationError::ValidationOutcome(
+            ValidationOutcome::PreflightResponseSignature((*preflight_response).clone()),
+        ))
+    }
+}
+
+/// Verify the difference between the end and start time is larger than the session header time offset.
+pub fn check_countersigning_session_times(
+    session_times: &CounterSigningSessionTimes,
+) -> SysValidationResult<()> {
+    let times_are_valid = &Timestamp(0, 0) < session_times.start()
```

Comment:
Why does the start of the session need to be < 0?

Reply:
@freesig it needs to be in "the future", `0` is in the past, it's also a common bug to have a `0` time so i thought i'd just check this

i suppose technically it is not a required check